src/js/components/pages/tag-editor.jsx

```diff
@@ -0,0 +1,217 @@
+        tags: null,
+        suggestions: ["none"],
+        mediaObjectSelected: false
+      };
+    },
+
+    componentDidMount: function () {
+        //subscribe to changes
+        SceneStore.addChangeListener(this._onChange);
+        GridStore.addChangeListener(this._onFocusChange)
+    },
+
+    componentWillUnmount: function () {
+        if (saveTimeout) {
+            //this.saveToScene()
+            clearTimeout(saveTimeout);
+        }
+        SceneStore.removeChangeListener(this._onChange);
+        GridStore.removeChangeListener(this._onFocusChange)
+    },
+
+    componentDidUpdate: function() {
+      //schedules a save in 1 second. Avoids to many saves
+      if (this.state.shouldSave) {
+        this.setState({shouldSave: false})
+        this.setSave()
+      }
+    },
+
+//autosave functions
+
+    setSave() {
+        console.log("TagEditor: Save Scheduled")
+        if (saveTimeout) {
+            clearTimeout(saveTimeout);
+        }
+        saveTimeout = setTimeout(this.saveToScene, 3000);
+    },
+
+    saveToScene() {
+
+      var tags = [];
+
+      if (this.state.tags.length > 0) {
+        for (var i = 0; i < this.state.tags.length; i++) {
+          tags.push(this.state.tags[i].text)
+        }
+      }
+
+      var csv = tags.toString();
+      console.log(csv);
+
+      try {
+          var scene = this.state.scene
+          scene.scene[this.state.focusedMediaObject].tags = csv;
+          SceneActions.updateScene(scene)
+          console.log("TagEditor: Changes saved");
+      } catch (error) {
+          console.log("TagEditor: Failed to save changes:", error)
+      }
+
+    },
+
+//change handlers
+
+    componentWillReceiveProps: function (nextProps) {
+      var scene = SceneStore.getScene(this.props._id)
+      var focusedMediaObject = GridStore.getFocusedMediaObject();
+
+      var mediaObject = null;
+      var tags = [];
+
+      if (focusedMediaObject != null) {
+
+      //have to use try catch because focusedMediaObject can be out of bounds
+      try {
+        mediaObject = scene.scene[focusedMediaObject]
+      } catch (error) {
+        console.log("TagEditor: failed to read media object", error);
+      }
+
+      }
+
+      if (mediaObject != null) {
+
+
+      //get the tags from the scene object
+      var mediaObjectTags = mediaObject.tags.split(",");
+
+      //get rid of spaces and format for tag editor
+      if (mediaObjectTags.length >= 1) {
+        for (var i = 0; i < mediaObjectTags.length; i++) {
+          mediaObjectTags[i] = mediaObjectTags[i].trim();
+
+          //add tags as long as they are not empty
+          if (mediaObjectTags[i] != "") {
+            tags.push({id: i, text: mediaObjectTags[i].trim()})
+          }
+
+        }
+        }
+      }
+      //update display
+      this.setState({scene: scene, suggestions: this.getSceneTags(scene), focusedMediaObject: focusedMediaObject, tags: tags});
+
+      },
+
+    _onChange: function () {
+
```

Comment:
We may wish to update this components state here.

If the scene store emits a change, the Scene JSON has likely changed.

I'll do some more investigation.

Reply:
This really boils down to - how much interoperability do we wish to support with authors potentially writing tags in either the Scene JSON editor or this component.  

This would also impact multiple users editing the same file, as updates might come from other machines rather than just the Monaco Editor.  We don't really have transactional support for multiple users, but the use of websockets has allowed us to keep the same scene updated on multiple screens.

Reply:
I did everything on will receive props as this seems to fire on any scene or focus change within the grid. I'm not 100% familiar with the update pattern though so will take another look and see if anything is being missed with doing it this way

Reply:
Yep so your receive props is good for the scene change, but for the same scene, if you edit the tag field of a media object through JSON, this component will not reflect that change.

It's a minor fix for sure 👍

Reply:
The _onChange will only be helpful for if the same scene is updated from a different component.

Reply:
Ok, makes sense to add that in then, I should be able to push that and a couple of other additions tomorrow :)

Reply:
Awesome!

I'll get some Semantic queries written to add another source of suggestions to this feature.  We aren't 100% there to be able to add this straight in as I still have a little bit more work left on it as a whole.  The main reason for this is allow suggestions to come from levels higher than a scene.  The query can look at media objects, themes, tags.  I'll demo this next time we are in the office.

We should also consider using other APIs like Thesaurus API too.  Something like this would really help.  

Another consideration is, at the project level, set vocabulary could be provided to aid the tagging process.  This lives partly with the Semantic modelling but it has a huge overlap with this authoring workflow.